catt/controllers.py

```diff
@@ -35,8 +41,37 @@ def get_chromecast(device_name):
         return devices[0]
 
 
-def human_time(seconds):
-    return time.strftime("%H:%M:%S", time.gmtime(seconds))
+def setup_cast(device_name, video_url=None, prep=None):
+    cache = Cache()
+    cached_ip = cache.get(device_name)
+
+    try:
+        if not cached_ip:
+            raise ValueError
+        cast = pychromecast.Chromecast(cached_ip)
+    except (pychromecast.error.ChromecastConnectionError, ValueError):
+        cast = get_chromecast(device_name)
+        cache.set(cast.name, cast.host)
+    cast.wait()
+
+    if video_url:
+        cc_info = (cast.device.manufacturer, cast.model_name)
+        stream = StreamInfo(video_url, model=cc_info, host=cast.host)
+        try:
+            if stream.is_local_file:
+                raise ValueError
+            app_id = APP_IDS[stream.extractor]
+        except (KeyError, ValueError):
+            app_id = DEFAULT_APP_ID
+    else:
+        stream = None
+        app_id = cast.app_id
+
+    if app_id == APP_IDS["youtube"] and cast.cast_type != "audio":
+        controller = YoutubeCastController(cast, "YouTube", APP_IDS["youtube"], prep=prep)
+    else:
+        controller = DefaultCastController(cast, "default", DEFAULT_APP_ID, prep=prep)
+    return (controller, stream) if stream else controller
```

Comment:
I know this is a bit creative, but it seemed like the least weird solution. At one point I considered putting the `stream` object inside the controller object, but decided against it, as I think this solution makes the `CastController` classes themselves more concise. I'd also like to preserve the explicit nature of current `cli.py`.

Reply:
Hmm, yeah. It makes sense to me, but I'm wondering whether this is a "controller" matter or an "app" matter, ie whether we should have different application classes that get passed to the controller, rather than different controller classes.

E.g. `controller = DefaultCastController(cast, "default", DEFAULT_APP_ID, prep=prep, app=YoutubeApp())`. I'm afraid I don't remember enough about the architecture of the Chromecast to definitely say whether this is a better approach, I'm just wondering whether the catt controller should be decoupled from the code that controls app that's running on the Chromecast.

I'll defer to you on this, as your knowledge is more current.

Reply:
> I'm just wondering whether the catt controller should be decoupled from the code that controls app that's running on the Chromecast.

I'm not sure I understand what you are saying. My main goal was to throw uniform controller objects back to the cli, regardless of what app the user wants to use/control. How can I achieve that if controller and app-control is still decoupled?

Reply:
I meant that the controller will accept its app-specific code as an argument, and the main Controller class will be app-agnostic. As I said, though, I don't really have an intuition of the high-level app flow on the Chromecast, so I'm probably way off. Can you explain a bit how e.g. the YouTube app differs from the default video controller app?